File: src/main/java/com/mallang/auth/domain/oauth/OauthMemberClientComposite.java
package com.mallang.auth.domain.oauth;

import static java.util.function.Function.identity;
import static java.util.stream.Collectors.toMap;

import com.mallang.auth.domain.OauthId.OauthServerType;
import com.mallang.auth.domain.OauthMember;
import com.mallang.auth.exception.UnsupportedOauthTypeException;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

@Profile("!test")
@Component
public class OauthMemberClientComposite {

    private final Map<OauthServerType, OauthMemberClient> mapping;

    public OauthMemberClientComposite(Set<OauthMemberClient> clients) {
        mapping = clients.stream()
                .collect(toMap(OauthMemberClient::supportServer, identity()));
    }

    public OauthMember fetch(OauthServerType oauthServerType, String authCode) {
        return getClient(oauthServerType).fetch(authCode);
    }

    private OauthMemberClient getClient(OauthServerType oauthServerType) {
        return Optional.ofNullable(mapping.get(oauthServerType))
                .orElseThrow(UnsupportedOauthTypeException::new);
    }
}

File: src/main/kotlin/com/mallang/auth/domain/oauth/OauthMemberClientComposite.kt
package com.mallang.auth.domain.oauth

import com.mallang.auth.domain.Member
import com.mallang.auth.domain.OauthServerType
import com.mallang.auth.exception.UnsupportedOauthTypeException
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import java.util.function.Function
import java.util.stream.Collectors

@Profile("!test")
@Component
class OauthMemberClientComposite(clients: Set<OauthMemberClient>) {

    private val mapping: Map<OauthServerType, OauthMemberClient> = clients.stream()
            .collect(Collectors.toMap({ it.supportServer() }, Function.identity()))

    fun fetch(oauthServerType: OauthServerType, authCode: String): Member {
        return getClient(oauthServerType).fetch(authCode)
    }

    private fun getClient(oauthServerType: OauthServerType): OauthMemberClient {
        return mapping[oauthServerType] ?: throw UnsupportedOauthTypeException()
    }
}
